### DISCUSSION TITLE
Unable to use TAB key or move directly from filed to another in the PDF form

### DISCUSSION THREAD

**Discussion by fadel2ab** [stephanrauh/ngx-extended-pdf-viewer#1708] [Q&A]
Hello all,

I try to use the latest version 16.2.0  instead of version 13.0.0 but I have a problem when I fill in the field and click on the TAB key to move to the next field. I notice that the value of the previous field becomes "null".

<img width="817" alt="image" src="https://user-images.githubusercontent.com/44398197/224048173-fc4aeef3-abff-4289-b01a-22a3d30232b2.png">

**It work if I fill the field and I click outside of form** :

<img width="746" alt="image" src="https://user-images.githubusercontent.com/44398197/224049195-08ec1a07-76d6-494c-bff8-0452a060af41.png">

**Can someone help me to know why and if there is an error in the PDF creation or there is a real bug ?**

URL to get the PDF : https://impotsdirects.public.lu/dam-assets/fr/formulaires/fiches_d_impot/2022/160F-2022.pdf 

Best regards,

Fadel

**Comment by stephanrauh** [comment]
It smells like a bug. Can you help us reproducing the problem?

- https://pdfviewer.net/extended-pdf-viewer/scripting uses the same file (or so I guess). Can you observe the same bug there?
- What is the JSON object in the screenshot? Are you using two-way binding, or is it an internal data structure of pdf.js?

**Reply by fadel2ab** [reply]
Yes sure.

I use two-way binding based on the property [(formData)]="formData" AND I handle the data with the below function : 

```
set formData(data: {[fieldName: string]: string | boolean | number}) {
    console.log(data)
    this._formData = data;
    this.acroformChanged.emit();
  }
```

The answers about your two points:

- Yes I have the same problem using this PDF  https://pdfviewer.net/extended-pdf-viewer/scripting
- The JSON object in the screen screenshot is the json data collected by above function formData().


I will be available for any further information.

NOTE:  

- It worked if I use this PDF https://pdfviewer.net/extended-pdf-viewer/forms . 
- I don’t have this problem when using the PDF [in first comment](https://pdfviewer.net/extended-pdf-viewer/scripting) over the version 13.0.0.

**Reply by stephanrauh** [reply]
Merci  beaucoup! I believe your input is going to help me to hunt down the bug (if there's a bug). I can't promise anything because I also have other obligations during the next couple of weeks, and  I'm drowning in a flood of new tickets.  Nonetheless, I plan to spend some time with your issue this week-end. Feel free to nudge me by the end of next week if you don't hear anything from me.

A tout à l'heur (ou à jour ou semaine?) :) 
Stephan

**Reply by fadel2ab** [reply]
Thank you for your responsiveness.

I have another small question regarding the new pop-up that comes out when I click F5 (refresh) which blocks seleniums tests. 

<img width="763" alt="image" src="https://user-images.githubusercontent.com/44398197/224279532-c85e2890-0f8c-469d-a059-68fc45c2d7c1.png">


I know that is possible to select the "Reload" button via CSS or disable this pop-up via the below code : 

`window.addEventListener('beforeunload', function (event) {
     event.stopImmediatePropagation();
});`


I would like to know if it is possible to have an option to disable this pop-up if is related to PDFviewer  ?

A tout à l'heure :) 
Fadel

**Reply by fadel2ab** [reply]
Hello  Stephan,

I come back to you in order to know if you had time to look at this ticket ?

Thank you for your understanding.

Fadel

**Reply by stephanrauh** [reply]
Oops. I knew there was something I wanted to do ASAP. The next week is going to be packed to the rim again - so feel free to nudge me again if you don't hear from me within the next seven days.

**Reply by stephanrauh** [reply]
I've managed to reproduce the issue.

**Reply by fadel2ab** [reply]
Great news 👍

**Reply by fadel2ab** [reply]
I had the same error with versions 16, 17 and 18. I would like to know if the problem has been resolved ?

**Reply by stephanrauh** [reply]
If it's a bug, it's better to open a bug ticket, because I tend to forget the discussions all the time.

Nonetheless, I've debugged the problem with this particular document. In my case, the field value becomes "undefined", not "null". When I tab from the name field to the first name field, I observe many `(formDataChange)` events. I'd expect only a single event. And the `undefined` is generated in the JavaScript code that's embedded in the PDF file. So I suspect it's a problem with the JavaScript code or the JavaScript engine of pdf.js. Unfornuately, that's something I can't debug.
